test/field/detail/CMSFieldMapReader.cc

```diff
@@ -0,0 +1,71 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file CMSFieldMapReader.cc
+//---------------------------------------------------------------------------//
+#include "CMSFieldMapReader.hh"
+
+#include <fstream>
+#include <iomanip>
+
+#include "base/Assert.hh"
+#include "base/Macros.hh"
+#include "base/Types.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the reader using an environment variable to get the volume-based
+ * CMS magnetic field map data.
+ */
+CMSFieldMapReader::CMSFieldMapReader()
+{
+    file_name_ = std::getenv("USER_FIELD_MAP");
+    CELER_VALIDATE(file_name_, << "USER_FIELD_MAP is not defined");
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Read the CMS volume-based magnetic field map data.
+ */
+CMSFieldMapReader::result_type CMSFieldMapReader::operator()() const
+{
+    result_type result;
+
+    // Set field map parameters
+    result.params.num_grid_r = 900 + 1;
+    result.params.num_grid_z = 2 * 1600 + 1;
+    result.params.offset_z   = real_type{1600};
+
+    // Store field values from the map file
+    std::ifstream ifile_(file_name_,
+                         std::ios::in | std::ios::binary | std::ios::ate);
+
+    if (ifile_.is_open())
+    {
```

Comment:
I know this is just "test" code now but it could evolve to be a demo app or the like 😉  This should probably be like the other file readers, where you `CELER_VALIDATE(ifile, << "Failed to open " << file_name_);`.